pennylane/decomposition/decomposition_graph.py

```diff
@@ -202,56 +218,63 @@ def _recursively_add_op_node(self, op_node: CompressedResourceOp) -> int:
+                # If an operator decomposes to nothing (e.g., a Hadamard raised to a
+                # power of 2), we must still connect something to this decomposition
+                # node so that it is accounted for.
+                self._graph.add_edge(self._start, d_node_idx, 0)
+            for op in decomp_resource.gate_counts:
+                op_node_idx = self._add_op_node(op)
+                self._graph.add_edge(op_node_idx, d_node_idx, (op_node_idx, d_node_idx))
+            self._graph.add_edge(d_node_idx, op_idx, 0)
         except DecompositionNotApplicable:
             pass  # ignore decompositions that are not applicable to the given op params.
 
-    def _get_adjoint_decompositions(self, op: CompressedResourceOp) -> list[DecompositionRule]:
-        """Retrieves a list of decomposition rules for an adjoint operator."""
+    def _get_adjoint_decompositions(self, op) -> list[DecompositionRule]:
+        """Gets the decomposition rules for the adjoint of an operator."""
 
         base_class, base_params = op.params["base_class"], op.params["base_params"]
 
+        # Special case: adjoint of an adjoint cancels out
         if issubclass(base_class, qml.ops.Adjoint):
             return [cancel_adjoint]
 
-        if (
-            issubclass(base_class, qml.ops.Pow)
-            and base_params["base_class"] in same_type_adjoint_ops()
-        ):
-            return [adjoint_pow_decomp]
-
-        if base_class in same_type_adjoint_ops():
-            return [same_type_adjoint_decomp]
-
-        if (
-            issubclass(base_class, qml.ops.Controlled)
-            and base_params["base_class"] in same_type_adjoint_ops()
-        ):
-            return [adjoint_controlled_decomp]
-
-        base_rep = resource_rep(base_class, **base_params)
-        return [AdjointDecomp(base_rule) for base_rule in self._get_decompositions(base_rep)]
+        # General case: apply adjoint to each of the base op's decomposition rules.
+        base = resource_rep(base_class, **base_params)
+        return [make_adjoint_decomp(base_decomp) for base_decomp in self._get_decompositions(base)]
 
     @staticmethod
-    def _get_pow_decompositions(op: CompressedResourceOp) -> list[DecompositionRule]:
-        """Retrieves a list of decomposition rules for a power operator."""
+    def _get_pow_decompositions(op) -> list[DecompositionRule]:
+        """Gets the decomposition rules for the power of an operator."""
 
         base_class = op.params["base_class"]
 
+        # Special case: power of zero
+        if op.params["z"] == 0:
+            return [null_decomp]
+
+        if op.params["z"] == 1:
+            return [decompose_to_base]
+
+        # Special case: power of a power
         if issubclass(base_class, qml.ops.Pow):
             return [merge_powers]
 
+        # Special case: power of an adjoint
+        if issubclass(base_class, qml.ops.Adjoint):
+            return [flip_pow_adjoint]
```

Comment:
We actually apply the `pow` first, then `adjoint`, because typically `adjoint` does not turn the operator into a different type, but `pow` might. For example, `pow(SX, 2)` is a `PauliX`. If we apply the `adjoint` on the `SX` first, `SX` gets decomposed, while we could've just had `PauliX`.